File: javastory.budgetsh.stage3/src/javastory/budgetsh/stage3/budget/ui/console/CashBookConsole.java
package javastory.budgetsh.stage3.budget.ui.console;

import javastory.budgetsh.stage3.budget.entity.budget.CashBook;
import javastory.budgetsh.stage3.budget.entity.travel.Travel;
import javastory.budgetsh.stage3.budget.logic.ServiceLogicLycler;
import javastory.budgetsh.stage3.budget.service.CashBookService;
import javastory.budgetsh.stage3.budget.share.DatePair;
import javastory.budgetsh.stage3.budget.share.IdName;
import javastory.budgetsh.stage3.budget.share.Socialian;
import javastory.budgetsh.stage3.budget.util.ConsoleUtil;
import javastory.budgetsh.stage3.budget.util.Narrator;
import javastory.budgetsh.stage3.budget.util.TalkingAt;
import javastory.budgetsh.stage3.club.entity.club.TravelClub;

public class CashBookConsole {
	//
	private ConsoleUtil consoleUtil;
	private Narrator narrator;

	private CashBookService cashbookService;

	public CashBookConsole() {
		//
		narrator = new Narrator(this, TalkingAt.Left);
		consoleUtil = new ConsoleUtil(narrator);

		cashbookService = ServiceLogicLycler.shareInstance().createCashBookService();
	}

	public CashBook register(TravelClub travelClub) {
		//mofify
		CashBook cashbook = null;
		Travel travel;
		Socialian socialian;

		while (true) {
			//
			String bankAccount = consoleUtil.getValueOf("\n Cashbook Bank Account(0.cashbook menu)");
			if (bankAccount.equals("0") || bankAccount.equals("") || bankAccount == null) {
				break;
			}
			if (cashbookService.exist(bankAccount)) {
				narrator.sayln("already exist this Account. -->" + bankAccount);
				continue;
			}

			String clubName = travelClub.getName();
			String travelName = clubName+"001";
			travel = new Travel(travelName);
			
			travel.setClub(new IdName(travel.getId(), clubName));

			String leaderSocialId = consoleUtil.getValueOf("\n SocialId");
			String leaderFirstName = consoleUtil.getValueOf("\n First Name");
			String leaderFamilyName = consoleUtil.getValueOf("\n FamilyName");
			String leaderEmail = consoleUtil.getValueOf("\n Leader Email");
			String leaderPhone = consoleUtil.getValueOf("\n Leader Phone Number");
			socialian = new Socialian(leaderSocialId, leaderFirstName, leaderFamilyName, leaderEmail);
			socialian.setPhone(leaderPhone);

			double monthlyDue = Double.parseDouble(consoleUtil.getValueOf("\n monthlyDue"));
			double budgetGoal = Double.parseDouble(consoleUtil.getValueOf("\n Budget Goal"));
			String startDate = consoleUtil.getValueOf("\n startDate(ex >2017-03-23)");
			String endDate = consoleUtil.getValueOf("\n End Date(ex > 2017-04-04)");

			cashbook = new CashBook(travel, monthlyDue, budgetGoal, new DatePair(startDate, endDate));
			cashbook.setBankAccount(bankAccount);
			cashbook.setClub(travel.getClub());
			cashbook.setLeader(socialian);

			cashbookService.regist(cashbook);
			narrator.sayln("Regist -->" + cashbook);
			break;
		}
		return cashbook;

	}

	public CashBook find() {
		//
		CashBook cashbook = null;

		String account = consoleUtil.getValueOf("\n Cashbook Account(0.cashbook Menu)");
		if (account.equals("0") || account.equals("") || account == null) {
			return null;
		}

		if (cashbookService.exist(account)) {
			cashbook = cashbookService.retrieve(account);
			narrator.sayln("\n > Found Cashbook" + cashbook);
			return cashbook;
		} else {
			narrator.sayln(" > No cashbook in storage -->" + account);
		}

		return cashbook;
	}

	public CashBook modify() {
		//
		CashBook foundCashbook = find();
		String temp;

		double monthlyDue;
		double budgetGoal;
		String endDate;

		if (foundCashbook != null) {
			if (!(temp = consoleUtil.getValueOf("\n >monthly Due (Enter. no change):")).equals("")) {
				monthlyDue = Double.parseDouble(temp);
			} else {
				monthlyDue = foundCashbook.getMonthlyDue();
			}
			if (!(temp = consoleUtil.getValueOf("\n >budgetGoal (Enter. no change):")).equals("")) {
				budgetGoal = Double.parseDouble(temp);
			} else {
				budgetGoal = foundCashbook.getBudgetGoal();
			}

			endDate = consoleUtil.getValueOf("\n > End Date (Enter. no change): ");
			if (endDate.equals("")) {
				endDate = foundCashbook.getTerm().getEndDate();
			}

			foundCashbook.setTerm(new DatePair(foundCashbook.getTerm().getStartDate(), endDate));
			foundCashbook.setBudgetGoal(budgetGoal);
			foundCashbook.setMonthlyDue(monthlyDue);

			cashbookService.update(foundCashbook);

		} else {
			narrator.sayln("\n >No cashbook in the storage");
		}

		return foundCashbook;
	}

	public void remove() {
		//
		CashBook foundCashbook;

		String bankAccount = consoleUtil.getValueOf("\n Cashbook Bank Account(0.cashbook menu)");
		if (bankAccount.equals("0") || bankAccount.equals("") || bankAccount == null) {
			return;
		}
		if (cashbookService.exist(bankAccount)) {
			foundCashbook = cashbookService.retrieve(bankAccount);
			cashbookService.remove(foundCashbook);
			narrator.sayln("removed -->" + foundCashbook.getBankAccount());
		}
	}

}

File: javastory.budgetsh.stage5.1/src/ui/console/CashBookConsole.js
const CashbookService = require('../../logic/CashBookService.js');
const Travel = require('../../domain/travel/Travel.js');
const IdName = require('../../domain/share/IdName.js');
const Socialian = require('../../domain/share/Socialian.js');
const CashBook = require('../../domain/budget/CashBook.js');
const cashbookService = new CashbookService();
const DatePair = require('../../domain/share/DatePair.js');
const MainMenu = require('../menu/MainMenu.js');
var co = require('co');
var prompt = require('co-prompt');
var program = require('commander');


class CashBookConsole {
	constructor() {
		this.cashbook = null;
		this.travel = null;
		this.socialian = null;
		this.bankAccount = '';
		this.travelName = '';
		this.clubName = '';
		this.leaderSocialId = '';
		this.leaderFirstName = '';
		this.leaderFamilyName = '';
		this.leaderEmail = '';
		this.leaderPhone = '';
		this.monthlyDue = 0.0;
		this.budgetGoal = 0.0;
		this.startDate = '';
		this.endDate = '';

	}

	register(previousMenu) {

		co(function*() {
			this.bankAccount = yield prompt('\n Cashbook Bank Account: ');
			if (cashbookService.exist(bankAccount)) {
				console.log("already exist this Account. -->" + bankAccount);
				register();
			}
			this.travelName = yield prompt('\n Travel Name: ');
			this.clubName = yield prompt('\n club Name: ');
			this.travel = new Travel(this.travelName);
			this.travel.club = new IdName(this.travel.id, this.clubName);

			this.leaderSocialId = yield prompt('\n SocialId: ');
			this.leaderFirstName = yield prompt('\n First Name: ');
			this.leaderFamilyName = yield prompt('\n FamilyName: ');
			this.leaderEmail = yield prompt('\n Leader Email: ');
			this.leaderPhone = yield prompt('\n Leader Phone: ');
			this.socialian = new Socialian(this.leaderSocialId, this.leaderFirstName, this.leaderFamilyName, this.leaderEmail);
			this.socialian.phone = this.leaderPhone;

			this.monthlyDue = Number(yield prompt('\n monthlyDue: '));
			this.budgetGoal = Number(yield prompt('\n Budget Goal: '));
			this.startDate = yield prompt('\n startDate(ex>2017-03-23)');
			this.endDate = yield prompt('\n End Date(ex>2018-01-10)');

			this.cashbook = new CashBook(this.travel, this.monthlyDue, this.budgetGoal, new DatePair(this.startDate, this.endDate));
			this.cashbook.bankAccount = bankAccount;
			this.cashbook.club = this.travel.club;
			this.cashbook.leader = this.socialian;

			cashbookService.regist(this.cashbook);
			
			console.log('Regist -->');
			this.cashbook.cashbook();
			//previousMenu.show();
			MainMenu.selectCashbookMenu();
			return this.cashbook;
		});
	}

	find(previousMenu) {
		co(function*() {
			this.bankAccount = yield prompt('\n Cashbook Account: ');
			if (cashbookService.exist(this.bankAccount)) {
				this.cashbook = cashbookService.retrieve(this.bankAccount);
				this.cashbook.cashbook();
				//previousMenu.show();
				MainMenu.selectCashbookMenu();
				return this.cashbook;
			} else {
				console.log('> No Cashbook in storage -->' + this.bankAccount);
			}
			//previousMenu.show();
			MainMenu.selectCashbookMenu();
			return this.cashbook;
		});
	}

	modify(previousMenu) {
		var foundCashbook = null;

		co(function*() {
				this.bankAccount = yield prompt('\n Cashbook Account: ');
				foundCashbook = cashbookService.retrieve(this.bankAccount);
				if(foundCashbook == null){
					console.log('> No Cashbook in storage.');
					//previousMenu.show();
					MainMenu.selectCashbookMenu();
					return;
				}
				this.monthlyDue = yield prompt('\n >monthly Due: ');
				if (this.monthlyDue == '')
					this.budgetGoal = yield prompt('\n >BudgetGoal: ');
				this.endDate = yield prompt('\n >End Date: ');

				this.term = new DatePair(foundCashbook.term.startDate, this.endDate);
				foundCashbook.term = this.term;
				foundCashbook.budgetGoal = this.budgetGoal;
				foundCashbook.monthlyDue = this.monthlyDue;

				cashbookService.update(foundCashbook);
				console.log('Updated --> ');
				foundCashbook.cashbook();
				//previousMenu.show();
				MainMenu.selectCashbookMenu();
		});
	}

	remove(previousMenu) {
		var foundCashbook = null;

		co(function*() {
			this.bankAccount = yield prompt('\n Cashbook BankAccount: ');
			if (cashbookService.exist(bankAccount)) {
				foundCashbook = cashbookService.retrieve(this.bankAccount);
				cashbookService.remove(foundCashbook);
				console.log('removed --> ');
				foundCashbook.cashbook();
				//previousMenu.show();
				MainMenu.selectCashbookMenu();
			}
		});
	}
}
cashBookConsole = new CashBookConsole();



module.exports = CashBookConsole;